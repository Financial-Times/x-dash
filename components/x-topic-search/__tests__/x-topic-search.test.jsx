const fetchMock = require('fetch-mock')
const { h } = require('@financial-times/x-engine')
const { mount } = require('@financial-times/x-test-utils/enzyme')
const { TopicSearch } = require('../')

const minSearchLength = 3
const maxSuggestions = 3
const apiUrl = 'api-url'
const FOLLOWED_TOPIC_ID1 = 'Cat-House-id'
const FOLLOWED_TOPIC_ID2 = 'Cat-Food-id'
const UNFOLLOWED_TOPIC_ID1 = 'Cat-Toys-id'

describe('x-topic-search', () => {
	const buildSearchUrl = (term) => `${apiUrl}?count=${maxSuggestions}&partial=${term}`
	const enterSearchTerm = (searchTerm) => {
		target.find('input').simulate('input', { target: { value: searchTerm } })

		return new Promise((resolve) => {
			setTimeout(resolve, 400)
		})
	}
	let target

	beforeEach(() => {
		const props = {
			minSearchLength,
			maxSuggestions,
			apiUrl,
			followedTopicIds: [FOLLOWED_TOPIC_ID1, FOLLOWED_TOPIC_ID2]
		}
		target = mount(<TopicSearch {...props} />)
	})

	afterEach(() => {
		fetchMock.reset()
	})

	describe('initial rendering', () => {
		it('should render with input box', () => {
			expect(target.find('input').exists()).toBe(true)
		})

		it('should not display result container', () => {
			expect(target.render().children('div')).toHaveLength(1)
		})
	})

	describe('when input receives focus', () => {
		it('selects the text in the input', () => {
			const selectMock = jest.fn()
			const inputBox = target.find('input')

			inputBox.simulate('blur')
			inputBox.simulate('focus', { target: { select: selectMock } })

			expect(selectMock).toHaveBeenCalledTimes(1)
		})
	})

	describe('given inputted text is shorter than minSearchLength', () => {
		const apiUrlWithResults = buildSearchUrl('a')

		beforeEach(() => {
			fetchMock.get(apiUrlWithResults, [])
			return enterSearchTerm('a')
		})

		it('does not make a request to the api or render any result', () => {
			expect(fetchMock.called(apiUrlWithResults)).toBe(false)
			expect(target.render().children('div')).toHaveLength(1)
		})
	})

	describe('given searchTerm which has some topic suggestions to follow', () => {
		const apiUrlWithResults = buildSearchUrl('Cat')
		const results = [
			{ id: FOLLOWED_TOPIC_ID1, prefLabel: 'Cat House', url: 'Cat-House-url' },
			{ id: FOLLOWED_TOPIC_ID2, prefLabel: 'Cat Food', url: 'Cat-Food-url' },
			{ id: UNFOLLOWED_TOPIC_ID1, prefLabel: 'Cat Toys', url: 'Cat-Toys-url' }
		]

		beforeEach(() => {
			fetchMock.get(apiUrlWithResults, results)
			return enterSearchTerm('Cat')
		})

		it('requests the topic suggestions with count set to maxSuggestions', () => {
			expect(fetchMock.called(apiUrlWithResults)).toBe(true)
		})

		it('renders no more than the max number of suggestions', () => {
			expect(target.render().children('div')).toHaveLength(2)
			expect(target.render().find('li')).toHaveLength(maxSuggestions)
		})

		it('renders links and follow buttons for each suggestion', () => {
			const suggestionsList = target.render().find('li')

			results.forEach((topic, index) => {
				const suggestion = suggestionsList.eq(index)

				expect(suggestion.find('a').text()).toEqual(topic.prefLabel)
				expect(suggestion.find('a').attr('href')).toEqual(topic.url)
				expect(suggestion.find('button').text()).toEqual(
					topic.id === UNFOLLOWED_TOPIC_ID1 ? 'Add to myFT' : 'Added'
				)
			})
		})
	})

	describe('given searchTerm which has no topic suggestions to follow', () => {
		const apiUrlNoResults = buildSearchUrl('Dog')

		beforeEach(() => {
			fetchMock.get(apiUrlNoResults, [])
			return enterSearchTerm('Dog')
		})

		it('requests from the api and renders the no matching topics message', () => {
			expect(fetchMock.called(apiUrlNoResults)).toBe(true)

			const resultContainer = target.render().children('div').eq(1)

			expect(resultContainer).toHaveLength(1)
			expect(resultContainer.find('h2').text()).toMatch('No topics matching')
		})
	})
})
