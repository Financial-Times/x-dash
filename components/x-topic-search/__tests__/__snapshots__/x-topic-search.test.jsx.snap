// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`x-topic-search given all topics include search term are followed should render followed topics name list 1`] = `
<div
  class="n-ui-hide-core TopicSearch_container__2VM2e"
>
  <h2
    class="o-normalise-visually-hidden"
  >
    Search for topics, authors, companies, or other areas of interest
  </h2>
  <label
    class="o-normalise-visually-hidden"
    for="topic-search-input"
  >
    Search and add topics
  </label>
  <div
    class="TopicSearch_input-wrapper__VeoLf"
  >
    <i
      class="TopicSearch_search-icon__2proG"
    />
    <input
      class="TopicSearch_input__1Iq7K"
      data-trackable="topic-search"
      id="topic-search-input"
      placeholder="Search and add topics"
      type="search"
    />
  </div>
  <div
    class="n-ui-hide-core TopicSearch_result-container__34uXy"
    data-component="topic-search"
  >
    <div
      aria-live="polite"
      class="TopicSearch_all-followed__wgnv9"
    >
      You already follow 
      <span>
        <b>
          abcc
        </b>
         
      </span>
      <span>
        and 
        <b>
          abccd
        </b>
      </span>
    </div>
  </div>
</div>
`;

exports[`x-topic-search given there are unfollowed topics include search term should render the unfollowed topics list with x-follow-button 1`] = `
<div
  class="n-ui-hide-core TopicSearch_container__2VM2e"
>
  <h2
    class="o-normalise-visually-hidden"
  >
    Search for topics, authors, companies, or other areas of interest
  </h2>
  <label
    class="o-normalise-visually-hidden"
    for="topic-search-input"
  >
    Search and add topics
  </label>
  <div
    class="TopicSearch_input-wrapper__VeoLf"
  >
    <i
      class="TopicSearch_search-icon__2proG"
    />
    <input
      class="TopicSearch_input__1Iq7K"
      data-trackable="topic-search"
      id="topic-search-input"
      placeholder="Search and add topics"
      type="search"
    />
  </div>
  <div
    class="n-ui-hide-core TopicSearch_result-container__34uXy"
    data-component="topic-search"
  >
    <ul
      aria-live="polite"
      class="TopicSearch_suggestions__2dZFW"
    >
      <li
        class="TopicSearch_suggestion__2z7vY"
        data-concept-id="TOPIC-1__id"
        data-trackable="myft-topic"
        data-trackable-meta="{\\"search-term\\":\\"abc\\"}"
      >
        <a
          class="TopicSearch_suggestion__name__3rwG-"
          data-trackable="topic-link"
          href="TOPIC-1__url"
        >
          TOPIC-1__name
        </a>
        <form
          action="/__myft/api/core/followed/concept/TOPIC-1__id?method=put"
          data-concept-id="TOPIC-1__id"
          data-myft-ui="follow"
          method="GET"
        >
          <input
            data-myft-csrf-token="true"
            name="token"
            type="hidden"
            value=""
          />
          <button
            aria-label="Add TOPIC-1__name to MyFT"
            aria-pressed="false"
            class="main_button__3Mk67 undefined"
            data-alternate-label="Remove TOPIC-1__name from MyFT"
            data-alternate-text="Added"
            data-concept-id="TOPIC-1__id"
            data-trackable="follow"
            title="Add TOPIC-1__name to MyFT"
            type="submit"
          >
            Add to myFT
          </button>
        </form>
      </li>
      <li
        class="TopicSearch_suggestion__2z7vY"
        data-concept-id="TOPIC-2__id"
        data-trackable="myft-topic"
        data-trackable-meta="{\\"search-term\\":\\"abc\\"}"
      >
        <a
          class="TopicSearch_suggestion__name__3rwG-"
          data-trackable="topic-link"
          href="TOPIC-2__url"
        >
          TOPIC-2__name
        </a>
        <form
          action="/__myft/api/core/followed/concept/TOPIC-2__id?method=put"
          data-concept-id="TOPIC-2__id"
          data-myft-ui="follow"
          method="GET"
        >
          <input
            data-myft-csrf-token="true"
            name="token"
            type="hidden"
            value=""
          />
          <button
            aria-label="Add TOPIC-2__name to MyFT"
            aria-pressed="false"
            class="main_button__3Mk67 undefined"
            data-alternate-label="Remove TOPIC-2__name from MyFT"
            data-alternate-text="Added"
            data-concept-id="TOPIC-2__id"
            data-trackable="follow"
            title="Add TOPIC-2__name to MyFT"
            type="submit"
          >
            Add to myFT
          </button>
        </form>
      </li>
    </ul>
  </div>
</div>
`;

exports[`x-topic-search given there is no topics include search term exist should render no topics message 1`] = `
<div
  class="n-ui-hide-core TopicSearch_container__2VM2e"
>
  <h2
    class="o-normalise-visually-hidden"
  >
    Search for topics, authors, companies, or other areas of interest
  </h2>
  <label
    class="o-normalise-visually-hidden"
    for="topic-search-input"
  >
    Search and add topics
  </label>
  <div
    class="TopicSearch_input-wrapper__VeoLf"
  >
    <i
      class="TopicSearch_search-icon__2proG"
    />
    <input
      class="TopicSearch_input__1Iq7K"
      data-trackable="topic-search"
      id="topic-search-input"
      placeholder="Search and add topics"
      type="search"
    />
  </div>
</div>
`;
